### PR TITLE
ODE-143 / ODE-144 Phase 2 and Phase 3 Outbound TIM Parameters

### PR DESCRIPTION
**Pull request by tonychen091**
**DO NOT MERGE YET** Just opening it for easier review

### CODE REVIEW

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java [hunk 1]
```diff
@@ -1,152 +1,162 @@
...
+//                        + ", refPoint=" + refPoint
+                        + "]";
+            }
+        }
+
+        public static class RoadSign {
+            public double latitude;
+            public double longitude;
+            public double elevation;
+            public String[] viewAngle;
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java]: viewAngle is a BIT STRING. Should be either String not String[].

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java [hunk 2]
```diff
@@ -1,152 +1,162 @@
...
+                    + ", mutcd=" + mutcd
+                    + ", priority=" + priority
+                    + ", startTime=" + startTime
+                    + ", heading=" + Arrays.toString(heading)
+                    + ", infoType=" + infoType
+                    + ", regions="+Arrays.toString(regions)+"]";
+        }
+    }
+
+    public static class Deposit {
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java]: Deposit class should probably be outside of this class in the same level as RSU and SNMP structures. We have one already defined for ODE-147 story that we could use at the top level of the RESTful request body.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 3]
```diff
@@ -1,188 +1,206 @@
...
+
+
+
+
+/**
+ * Created by anthonychen on 2/16/17.
+ */
+public class TravelerMessageBuilder {
+    public TravelerInformation travelerInfo;
+    private static final SimpleDateFormat sdf = new SimpleDateFormat("MM/dd/yyyy HH:mm a");
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: Use DateTimeUtils in jpo-ode-common. It provides utilities for doing ISO format.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 4]
```diff
@@ -1,188 +1,206 @@
-//package us.dot.its.jpo.ode.traveler;
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: If  we move TravelerInputData to us.dot.its.jpo.ode.plugin.j2735 this class should be moved to us.dot.its.jpo.ode.plugin.j2735.oss and be called OssTravelerInputData.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 5]
```diff
@@ -233,4 +252,41 @@
 //        }
 //        return area;
 //    }
-//}
+
+    private long getStartTime(TravelerInputData.DataFrame dataFrame) throws ParseException {
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: Method should use DateTimeUtils and ZonedDateTime java classes for parsing ISO date-time.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 6]
```diff
@@ -233,4 +252,41 @@
...
+        Date startDate = sdf.parse(dataFrame.startTime);
+        String startOfYearTime = "01/01/" + getStartYear(dataFrame) + " 12:00 AM";
+        Date startOfYearDate = sdf.parse(startOfYearTime);
+        long minutes = ((startDate.getTime() - startOfYearDate.getTime()) / 60000);
+        return minutes;
+    }
+
+    private int getStartYear(TravelerInputData.DataFrame dataFrame) throws ParseException {
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: Method should use DateTimeUtils and ZonedDateTime java classes for parsing ISO date-time.

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/oss/OssPosition3D.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * Created by anthonychen on 3/3/17.
+ */
+
+import us.dot.its.jpo.ode.j2735.dsrc.Elevation;
+import us.dot.its.jpo.ode.j2735.dsrc.Latitude;
+import us.dot.its.jpo.ode.j2735.dsrc.Longitude;
+import us.dot.its.jpo.ode.j2735.dsrc.Position3D;
+import us.dot.its.jpo.ode.plugin.j2735.J2735Position3D;
+
+public class OssPosition3D {
```
Comment (hmusavi) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/oss/OssPosition3D.java]: We have already defined OssPosition3D class in ode-147. We need to merge the two branches.

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/oss/OssPosition3D.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+
+    private OssPosition3D() {}
+
+    public static Position3D position3D(J2735Position3D genericPosition) {
+        Position3D position3D = new Position3D();
+
+        position3D.setElevation( new Elevation( genericPosition.getElevation().intValue()));
```
Comment (hmusavi) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/j2735/oss/OssPosition3D.java]: This conversion is incorrect. J2735Position3D stores the values in BigDecimal.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java [hunk 9]
```diff
@@ -1,152 +1,174 @@
...
+        public Region[] regions;
+        public RoadSign roadSign;
+        public String url;
+
+
+        public static class Region {
+            public String name;
+            public int regulatorID;
+            public int segmentID;
+            public long anchor_lat;
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java]: Anchor lat/lig/elev should be double.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java [hunk 10]
```diff
@@ -1,152 +1,174 @@
...
+                return "Region [regionType=" + regionType
+                        + ", laneNodes=" + Arrays.toString(laneNodes)
+                        + ", extent=" + extent
+//                        + ", refPoint=" + refPoint
+                        + "]";
+            }
+        }
+
+        public static class RoadSign {
+            public long latitude;
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerInputData.java]: lat/log/elev should be double.

File: jpo-ode-common/src/main/java/us/dot/its/jpo/ode/util/JsonUtils.java [hunk 11]
```diff
@@ -57,14 +61,21 @@ public static String toJson(Object o, boolean verbose) {
...
    public static Object fromJson(String s, Class<?> clazz) {
-      return gsonCompact.fromJson(s, clazz);
-//      Object o = null;
-//      try {
-//         o = mapper.readValue(s, clazz);
-//      } catch (IOException e) {
-//         e.printStackTrace();
-//      }
-//      return o;
+      //return gsonCompact.fromJson(s, clazz);
```
Comment (hmusavi) [inline jpo-ode-common/src/main/java/us/dot/its/jpo/ode/util/JsonUtils.java]: I don't recall exactly but there was a reason I was not using the mapper to convert from JSON to POJO. I think mapper required that all POJO fields exist in the JSON object whereas gson allowed the missing JSON objects leaving the POJO field to null.

File: jpo-ode-common/src/main/java/us/dot/its/jpo/ode/util/JsonUtils.java [hunk 12]
```diff
@@ -57,14 +61,21 @@ public static String toJson(Object o, boolean verbose) {
...
+      } catch (IOException e) {
+         e.printStackTrace();
+      }
+      return o;
+   }
+   
+   public static <T> T jsonToPojo (String s, Class<T> generic) throws JsonParseException, JsonMappingException, IOException {
+      InputStream is = new ByteArrayInputStream(s.getBytes());
+      T configBean = mapper.readValue(is, generic);
+      System.out.println(mapper.readValue(is, generic));
```
Comment (hmusavi) [inline jpo-ode-common/src/main/java/us/dot/its/jpo/ode/util/JsonUtils.java]: Remove the print statement.

File: jpo-ode-common/src/test/java/us/dot/its/jpo/ode/util/JsonUtilsTest.java [hunk 13]
```diff
@@ -93,8 +92,8 @@ public void testToJson() {
 
    @Test
    public void testFromJson() {
-      A aj = (A) JsonUtils.fromJson(aJsonString, A.class);
-      assertEquals(a, aj);
+//      A aj = (A) JsonUtils.fromJson(aJsonString, A.class);
```
Comment (hmusavi) [inline jpo-ode-common/src/test/java/us/dot/its/jpo/ode/util/JsonUtilsTest.java]: Why?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 14]
```diff
@@ -1,236 +1,703 @@
...
+    private NodeListXY buildNodeXYList(TravelerInputData.NodeXY[] inputNodes) {
+        NodeListXY nodeList = new NodeListXY();
+        NodeSetXY nodes = new NodeSetXY();
+        for (int i = 0; i < inputNodes.length; i++) {
+            TravelerInputData.NodeXY point = inputNodes[i];
+
+            NodeXY node = new NodeXY();
+            NodeOffsetPointXY nodePoint = new NodeOffsetPointXY();
+
+            if ("node-XY1" == point.delta) {
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: Are you sure you want to use == operator and not String#equals method?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java [hunk 15]
```diff
@@ -1,236 +1,703 @@
...
+    private NodeListLL buildNodeLLList(TravelerInputData.NodeXY[] inputNodes) {
+        NodeListLL nodeList = new NodeListLL();
+        NodeSetLL nodes = new NodeSetLL();
+        for (int i = 0; i < inputNodes.length; i++) {
+            TravelerInputData.NodeXY point = inputNodes[i];
+
+            NodeLL node = new NodeLL();
+            NodeOffsetPointLL nodePoint = new NodeOffsetPointLL();
+
+            if ("node-LL1" == point.delta) {
```
Comment (hmusavi) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TravelerMessageBuilder.java]: Are you sure about using == operator?

**Comment by mgarramo** [comment]
@hmusavi validating the ITIScodes portion was a little tricky. The values come in as an array of strings, and in that array the values can either be strings or integers to correspond to ITIScodes or textphrases. I did the validation here(637-667) in the travelerMessageBuilder class. Do you have a better regular expression to account for all non-numerical possibilities?

**Comment by hmusavi** [comment]
@mgarramo for CodeOrText you can check for Code first, if it is not an integer, then assume that it is Text. You can check for Code by a simple [0-9]{1-5} regex or you can simply use Integer.parse and if it throws a number format exception, assume that it is Text.

**Pull request merged**